Reference File Finding

rmbasilio opened the pull request:
This was the change I had made to fix the file searching. You don't have to use it, but I think you'll get more reliable and accurate results this way.

Basically, the main change is instead of using regex at the raw text file, I'm looking at the URL objects that are created in the Protobuffer data. Even though you were parsing the Protobuffer data, you were immediately converting it back into a long text file. Maybe it won't make a difference for most files, but it'll get you closer to what the data actually says.

There's a couple little optimizations in there too, like shortcuts where if you've matched at least one file, there's no need to check if there's more matches, and a couple checks to make sure you don't accidentally double check data or check empty data.

rmbasilio committed:
refactored to look directly at url objects instead of raw text

rmbasilio committed:
minor optimization

rmbasilio committed:
redo fix

rmbasilio commented:
and sorry it took forever to get this in, I spent forever trying to get it to stop saying the entire file had changed when it hadn't.

greyshirtguy commented:
Nice recursive enumeration of the fields! - I like it!

However, please note that the URL objects are defined using "string" datatypes for all three fields that are searched: (absolute_string, relative_path and path)

```
message URL {
  enum Platform {
    PLATFORM_UNKNOWN = 0;
    PLATFORM_MACOS = 1;
    PLATFORM_WIN32 = 2;
    PLATFORM_WEB = 3;
  }
  Platform platform = 3;
  oneof Storage {
    string absolute_string = 1;
    string relative_path = 2;
  }

  oneof RelativeFilePath {
    rv.data.URL.LocalRelativePath local = 4;
    rv.data.URL.ExternalRelativePath external = 5;
  }


  message LocalRelativePath {
    enum Root {
      ROOT_UNKNOWN = 0;
      ROOT_BOOT_VOLUME = 1;
      ROOT_USER_HOME = 2;
      ROOT_USER_DOCUMENTS = 3;
      ROOT_USER_DOWNLOADS = 4;
      ROOT_USER_MUSIC = 5;
      ROOT_USER_PICTURES = 6;
      ROOT_USER_VIDEOS = 7;
      ROOT_USER_DESKTOP = 11;
      ROOT_USER_APP_SUPPORT = 8;
      ROOT_SHARED = 9;
      ROOT_SHOW = 10;
      ROOT_CURRENT_RESOURCE = 12;
    }
    Root root = 1;
    string path = 2;

  }
```

Since those fields are (UTF8) strings, both methods result in the same string data. 
I'm afraid enumerating the fields and collecting the URLs and then later checking those string fields againts media file names won't avoid the issue - you would get the same result/issues with some Unicode characters.

The problem is that Unicode permits two ways of encoding/storing some (accented) chars and applications can freely choose between the two methods (composed/decomposed).  So two Unicode strings with exact same accented char may fail to match when they use different encoding methods.

The **solution is Unicode normalization before comparing the strings.**....(which you have added in one of your commits).
Enumerating the URLs won't help the issue _(although it does allow you to nicely collect GUIDs of media files and ensure to only add those once to the list!)_

I think that Arlin is already working on normalizing before comparing.... - so it will be interesting to see what he decides to do here.

rmbasilio commented:
Thanks greyshirt. I get what you're saying, but in practice I'm seeing it behave correctly this way as opposed to before, and I think I have a suspicion why. In particular, I'm looking at the local path (or `path`)

So in the original code, that `absolute_string` is actually a string, that's been explicitly encoded by the developers using "%dd" format, because this needs to be useable even if a human copy/pastes it into an address bar. This is why `absolute_string` always has the "%dd" encoding, no matter what, and we have to manually decode it: that's the actual text, there's nothing for the program to decode.(I mentioned this to Arlin, but I used to be one of the developers of Pro7 early on, this is the only reason I can speak with any confidence on this). 

But the local `path` is actually coming from the relative URL objects in Swift or C#. These aren't explicitly encoded by the developers, because it doesn't matter what it looks like in data, as long as when it comes back to ProPresenter, it comes back the same. It appears to be using "\nnn" encoding in the backend, at least in the examples I've tested.

The `path` string, if you pull it from the overall text file like was being done before, still has the "\nnn" encoding in there. But, if you pull it from the URL object, it has decoded the "\nnn" characters into readable text. I suspect that this is because when you pull it from the overall text file, you're not just reading values, you're reading an entire data structure, converted into string form, so not everything is user friendly yet. When you pull it from the URL object, this is the data completely decoded and ready to use, so the encoding translates correctly.

arlinsandbulte commented:
I still need to review, but my newest code does NOT normalize because I found out Windows filenames can coexist either way.  MacOS, on the other hand, would not let a filename that looks the same coexist in the same folder, at least using finder.  CLI might allow it.
It's Sunday morning now, so I gotta go.  But, I'll review and discuss more later.
Thanks all!

rmbasilio commented:
Also, because I realize it might not be clear, my findings of the \nnn being properly decoded were before the normalize function occurs, based on results during debugging. That was something I added on early on when I was still diagnosing, and left because it seemed like a good idea (commits are out of order because I forgot good git etiquette and didn't commit anything until I was completely done, and split up my edits into smaller chunks after the fact)

greyshirtguy commented:
My bad! - You are correct!
My assuption that the string types would make it through to the string representation of the protocol buffer un-encoded was **wrong**!

When asking Google Protocol buffer objects for their string representation they will be escaped as per:
https://chromium.googlesource.com/external/github.com/google/protobuf/+/HEAD/java/core/src/main/java/com/google/protobuf/TextFormatEscaper.java

When accessing the individual member of type sting - you will get the original (unescaped) Unicode string. These Unicode strings, still need to be normalized in some way before comparing (which converting to path objects may be doing - I havent checked)

I think I am just muddying up this conversation and will let you and Arlin take it from here!

arlinsandbulte commented:
Sorry for not following up sooner.  Been spending a lot of time on this project the last few days.  Church this morning.  Lunch with the fam.  Then, I crashed this afternoon.

@rmbasilio I tried this code on windows, but it swept up a bunch of media files that it should not have.
Not sure why yet.

Regarding the whole 'normalizing' thing:
Windows DOES distinguish between files with the same character, but different unicode byte code.  AND Pro7 does carry that through.  So, I am actually leaning towards NOT normalizing the comparisons.
For example, I can name two different media files in Windows:
'ñ.jpg' <-- the n~ here is a single byte character
'ñ.jpg' <-- the n~ here is a multi byte character
(Note: single/multi byte my not be exactly the correct term, but I hope you get the point.)
Those two files can co-exist in the same folder in Windows.  And, they can both be used in Pro7 without issue.
But, On Mac, I think those two file names are considered the same name and not allowed (at least Finder does not allow it).
I can't zip those files up either.  So, zip format does not allow it.
![image](https://user-images.githubusercontent.com/20544649/149683845-0c9f34bb-187c-469f-aace-e8d310f7462c.png)
![image](https://user-images.githubusercontent.com/20544649/149684122-76577a64-1fa6-4a1f-8553-108dacad5be5.png)

arlinsandbulte commented:
Overall, I think rmbasilio's approach is probably the 'right' way to do it.
And, I think it performs better too.

But, for now, I'm going to keep my original logic.
I'll spend more time going over @rmbasilio's code and hopefully fixing the erroneous sweeps.
Is that reasonable?

arlinsandbulte commented:
@rmbasilio 
Thanks for all of your effort on this.  I'm amazed how much work you did and how little time it took!
I'm going to work on a new release tonight with @The-Sizzler's GUI improvements.

arlinsandbulte commented:
@rmbasilio One more thing:
I'm using PyCharm's code style guidance to keep the code as clean as I can.
Keeps everything consistent and easy to spot my errors.
So, I just ask that any changes keep following that code style too.  Sorry for being so picky.

rmbasilio commented:
greyshirt, by all means keep commenting, I like your input! Your objection made sense, and I did some doublechecking to make sure it actually made a difference.

arlin, that all sounds good to me. It might make sense to add an OS check to the normalizing calls, or it might not be necessary - I haven't thought enough yet on the use cases. 
Regarding what it's not working on yet, it worked on the couple  cases I was looking at, but I didn't check all the files it pulled out yet to see if there were any false positives. I think until you're not getting any false positives, this should stay in a separate branch. 
And regarding code style, definitely, I agree! I actually spent maybe 2-3 hours yesterday just trying to get it to stop saying every line had changed. Something about line returns in Mac vs Windows, but the settings just refused to change, and then authentication issues with github, etc. But anyway, I tried to match the code style you were doing, but let me know if I missed anywhere. Sometimes autopilot takes over and I don't notice. I'll review Pycharm's rubrics so I know specifically what we're going for.

arlinsandbulte commented:
Mostly, I'm just fixing stuff when PyCharm flags it.
I forgot to attach this earlier:
![image](https://user-images.githubusercontent.com/20544649/149688264-a26f369f-5d33-426b-8e99-00da0145714f.png)

greyshirtguy commented:
@rmbasilio: You prob had different line-ending bytes in your text file! On Windows the default text file line endings are CR & LF. While on MacOS, the default text file line endings are just LF.  If you use a good text editor, you can see/chnage the line endings. eg: Visual Studio code shows it at the bottom right and you can click the LF/CRLF to change for your file.
![image](https://user-images.githubusercontent.com/1886513/149703349-f5446e4a-29b4-41b5-9abc-79ffd6880d97.png)

@arlinsandbulte: I'm not sure about choosing to intentionally NOT normalize....

I know you have tested and learned that Windows *lets* you create two different files using the different encoding (but having the same canonical name) - but according to Unicode that is probably incorrect behaviour:
[Unicode](http://unicode.org/faq/normalization.html)
![image](https://user-images.githubusercontent.com/1886513/149703786-64e3f7a0-6a4f-4bd3-9b3e-d89749d60c89.png)

When I read your description of the issue, I noted that normalization would make the two strings equal.  (Although I neglected to account for the need to decode the \ooo octal encoding that the string output of the Google Protocol Buffer uses for path)

![image](https://user-images.githubusercontent.com/1886513/149704029-3fe96ded-4ec7-450f-a515-318d27582563.png)

In the original issue:
You have a filename with the _composed_ Unicode bytes of the accented `ñ` (This is two bytes that represent that one char in UTF8).

But you also have Pro7 storing that same char in the (relative) `path` with the _decomposed_ Unicode bytes instead! (This is the n char plus two bytes that compose the tilde accent above it).  They will not compare as equal (even after you properly decode  the \ooo octal encoding in the Protocol Buffer string back to Unicode bytes). The "recommended" way that they can be compared  - is to convert to a common Unicode format (normalized).

If Pro7 is storing decomposed accented chars in relative path strings while the oringinal file name has a composed accented char - they will never match without normalizing both to some common encoding!  I might do some testing to check.

arlinsandbulte commented:
The code seems to be working well as-is.  So, I'm going to close this pull request without merging it in.
But, in the future, I'll definitely keep this method in mind and come back to it if/when other changes are needed.
Thanks!
